src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorFormattingPass.cs

```diff
@@ -86,11 +86,172 @@ private static IEnumerable<TextEdit> FormatRazor(FormattingContext context, Razo
                 TryFormatSingleLineDirective(context, edits, source, node);
+                TryFormatTransitionCodeBlock(context, edits, source, node);
             }
 
             return edits;
         }
 
+        private static void TryFormatTransitionCodeBlock(FormattingContext context, IList<TextEdit> edits, RazorSourceDocument source, SyntaxNode node)
+        {
+            Range? openBraceRange = null;
+            Range? codeRange = null;
+            Range? closeBraceRange = null;
+            SyntaxNode? openBraceNode = null;
+            SyntaxNode? codeNode = null;
+            SyntaxNode? closeBraceNode = null;
+
+            // TODO: Split the if's into their own methods
+
+            // complex situations like
+            // @{
+            //  public void Method(){
+            //      @(DateTime.Now)
+            //  }
+            // }
+            if (node is CSharpRazorBlockSyntax csharpRazorBlock &&
+                csharpRazorBlock.Parent is CSharpCodeBlockSyntax firstCodeBlock &&
+                csharpRazorBlock.Parent.Parent is CSharpCodeBlockSyntax secondCodeBlock)
+            {
+                var csharpRazorRange = csharpRazorBlock.GetRange(source);
+                var childPosition = 0;
+                for (var i = 0; i < secondCodeBlock.Children.Count; i++)
+                {
+                    var child = secondCodeBlock.Children[i];
+                    if (ReferenceEquals(child, firstCodeBlock))
+                    {
+                        childPosition = i;
+                        break;
+                    }
+                }
+
+                codeNode = csharpRazorBlock;
+                codeRange = codeNode.GetRangeWithoutWhitespace(source);
+                if (childPosition == 0)
+                {
+                    openBraceNode = codeNode;
+                    openBraceRange = codeRange;
+                }
+                else
+                {
+                    var child = secondCodeBlock.Children[childPosition - 1];
+                    openBraceNode = child;
+                    openBraceRange = child.GetRangeWithoutWhitespace(source);
+                }
+
+                if (childPosition == secondCodeBlock.Children.Count - 1)
+                {
+                    closeBraceRange = codeRange;
+                }
+                else
+                {
+                    var child = secondCodeBlock.Children[childPosition + 1];
+                    closeBraceNode = child;
+                    closeBraceRange = child.GetRangeWithoutWhitespace(source);
+                }
+            }
+            // void Method()
+            // {
+            //     <div></div>
+            // }
+            else if (node is MarkupBlockSyntax markupBlockNode &&
+                markupBlockNode.Parent is CSharpCodeBlockSyntax cSharpCodeBlock)
+            {
+                for (var i = 0; i < cSharpCodeBlock.Children.Count; i++)
```

Comment:
Seems like it would be just as good to use `int i = cSharpCodeBlock.Children.IndexOf(markupBlockNode)`, and would be more easy to understand I think. Should also make it so that all of the `openBraceNode`, `openBraceRange` etc. variables at the top don't have to be initialized to null.

I assume there aren't methods on the nodes to get their next or previous siblings?